### PR TITLE
Kalman

### PR DESCRIPTION
**Pull request by martindeegan**
Implemented the first stages of the EKF. Added a 3D visualizer.

### CODE REVIEW

File: copter/Cargo.toml [hunk 1]
```diff
@@ -4,25 +4,27 @@ version = "0.1.0"
...
-i2cdev-l3gd20 = "0.1.*"
-i2cdev-lsm303d = "0.1.*"
-i2cdev-lsm9ds0 = "0.1.*"
-i2cdev = "0.3.1"
-unbounded-gpsd = { path = "./unbounded-gpsd" }
-wifilocation = { path = "./wifilocation" }
-nalgebra = "0.13"
-typenum = "1.7.0"
-reqwest = "0.7.3"
+i2cdev-bmp180 = { path = "./local/i2cdev-sensors/i2cdev-bmp180" }
```
Comment (tadeegan) [inline copter/Cargo.toml:12]: Is this right?

File: copter/configuration/config_default.toml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+[flight.roll]
```
Comment (tadeegan) [inline copter/configuration/config_default.toml:1]: Whats the difference between toml and json configs?

File: copter/configurations/src/calibrations.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                println!("1: {:?}", e);
+                return Err(e.to_string());
+            }
+        };
+
+        let mut calibration_string = String::new();
+        match calibration_file.read_to_string(&mut calibration_string) {
+            Ok(_size) => {}
+            Err(e) => {
+                println!("2: {:?}", e);
```
Comment (tadeegan) [inline copter/configurations/src/calibrations.rs:81]: What does print 2 mean?

File: copter/configurations/src/calibrations.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+        calibration_file
+            .write_all(calibration_string.as_bytes())
+            .unwrap();
+        Ok(())
+    }
+}
+
+impl Default for Calibrations {
+    fn default() -> Calibrations {
```
Comment (tadeegan) [inline copter/configurations/src/calibrations.rs:119]: Where is this used?

File: copter/ekf_matlab/helper_functions.m [hunk 5]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (tadeegan) [inline copter/ekf_matlab/helper_functions.m]: Delete.

File: copter/ekf_matlab/ekf_derivation.m~ [hunk 6]
```diff
@@ -0,0 +1,137 @@
+% SafeFlight EKF derivation
```
Comment (tadeegan) [inline copter/ekf_matlab/ekf_derivation.m~]: Is this a deleted file?

File: copter/src/flight/altitude.rs [hunk 7]
```diff
@@ -26,7 +27,8 @@ impl Altitude {
     }
 
     pub fn get_mid_level(&mut self, current_altitude: f32, desired_altitude: f32, climb: Option<f32>, dt: f32) -> f32 {
-        let proportional = desired_altitude - current_altitude;
+        let mut proportional = desired_altitude - current_altitude;
+        proportional  = proportional * proportional * proportional;
```
Comment (tadeegan) [inline copter/src/flight/altitude.rs]: What is this?

**Review by tadeegan — APPROVED** [review]

File: copter/configuration/config_default.toml (hunk 2 above)
Comment (martindeegan) [reply copter/configuration/config_default.toml:1]: I switched everything to toml cuz it looks way nicer.

File: copter/configurations/src/calibrations.rs (hunk 4 above)
Comment (martindeegan) [reply copter/configurations/src/calibrations.rs:119]: If there is no calibration file.

File: copter/ekf_matlab/ekf_derivation.m~ (hunk 6 above)
Comment (martindeegan) [reply copter/ekf_matlab/ekf_derivation.m~]: Yep, Ill delete these and add them to the gitignore

File: copter/src/flight/altitude.rs (hunk 7 above)
Comment (martindeegan) [reply copter/src/flight/altitude.rs]: I'll delete this file.

File: copter/configurations/src/calibrations.rs (hunk 3 above)
Comment (martindeegan) [reply copter/configurations/src/calibrations.rs:81]: No clue, must be some left over debugging stuff.

**Pull request merged**